File: packages/oxlint-plugin-react-doctor/src/plugin/rules/r3f/three-texture-repeat-requires-wrapping.ts
import { defineRule } from "../../utils/define-rule.js";
import type { EsTreeNode } from "../../utils/es-tree-node.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";
import { findEnclosingFunction } from "../../utils/find-enclosing-function.js";
import { getRangeStart } from "../../utils/get-range-start.js";
import { getStaticPropertyName } from "../../utils/get-static-property-name.js";
import { isNodeOfType } from "../../utils/is-node-of-type.js";
import { resolveExpressionKey } from "../../utils/resolve-expression-key.js";
import type { RuleContext } from "../../utils/rule-context.js";
import { stripParenExpression } from "../../utils/strip-paren-expression.js";
import { DEFAULT_TEXTURE_REPEAT } from "./constants.js";
import { getStaticNumber } from "./utils/get-static-number.js";
import { getThreeConstructorName } from "./utils/get-three-constructor-name.js";
import { isThreeRepeatingWrapping } from "./utils/is-three-repeating-wrapping.js";

interface TextureWrappingFact {
  readonly axis: "x" | "y";
  readonly node: EsTreeNode;
  readonly owner: EsTreeNode | null;
  readonly textureKey: string;
}

const THREE_TEXTURE_CONSTRUCTOR_NAMES: ReadonlySet<string> = new Set([
  "CanvasTexture",
  "CompressedTexture",
  "Data3DTexture",
  "DataArrayTexture",
  "DataTexture",
  "DepthTexture",
  "FramebufferTexture",
  "Texture",
  "VideoTexture",
]);

export const threeTextureRepeatRequiresWrapping = defineRule({
  id: "three-texture-repeat-requires-wrapping",
  title: "Three.js texture repeat without repeat wrapping",
  category: "Correctness",
  severity: "warn",
  recommendation:
    "Set wrapS or wrapT to RepeatWrapping or MirroredRepeatWrapping before increasing repeat on that axis",
  create: (context: RuleContext) => {
    const wrappingFacts: TextureWrappingFact[] = [];
    return {
      AssignmentExpression(node: EsTreeNodeOfType<"AssignmentExpression">) {
        if (
          node.operator !== "=" ||
          !isNodeOfType(node.left, "MemberExpression") ||
          !isThreeRepeatingWrapping(node.right, context.scopes)
        ) {
          return;
        }
        const propertyName = getStaticPropertyName(node.left);
        if (propertyName !== "wrapS" && propertyName !== "wrapT") return;
        const textureKey = resolveExpressionKey(node.left.object, context);
        if (!textureKey) return;
        wrappingFacts.push({
          axis: propertyName === "wrapS" ? "x" : "y",
          node,
          owner: findEnclosingFunction(node),
          textureKey,
        });
      },
      CallExpression(node: EsTreeNodeOfType<"CallExpression">) {
        const callee = stripParenExpression(node.callee);
        if (!isNodeOfType(callee, "MemberExpression") || getStaticPropertyName(callee) !== "set") {
          return;
        }
        const repeatMember = stripParenExpression(callee.object);
        if (
          !isNodeOfType(repeatMember, "MemberExpression") ||
          getStaticPropertyName(repeatMember) !== "repeat" ||
          !THREE_TEXTURE_CONSTRUCTOR_NAMES.has(
            getThreeConstructorName(repeatMember.object, context.scopes) ?? "",
          )
        ) {
          return;
        }
        const textureKey = resolveExpressionKey(repeatMember.object, context);
        const repeatStart = getRangeStart(node);
        if (!textureKey || repeatStart === null) return;
        const owner = findEnclosingFunction(node);
        const axisValues = [
          { axis: "x" as const, expression: node.arguments[0] },
          { axis: "y" as const, expression: node.arguments[1] },
        ];
        for (const { axis, expression } of axisValues) {
          if (!expression || isNodeOfType(expression, "SpreadElement")) continue;
          const value = getStaticNumber(expression, context.scopes);
          if (value === null || value <= DEFAULT_TEXTURE_REPEAT) continue;
          const hasWrapping = wrappingFacts.some((fact) => {
            const wrappingStart = getRangeStart(fact.node);
            return (
              fact.axis === axis &&
              fact.textureKey === textureKey &&
              fact.owner === owner &&
              wrappingStart !== null &&
              wrappingStart < repeatStart
            );
          });
          if (hasWrapping) continue;
          context.report({
            node: expression,
            message: `texture.repeat.${axis} is greater than one, but the corresponding ${axis === "x" ? "wrapS" : "wrapT"} remains ClampToEdgeWrapping so the texture will not tile on that axis`,
          });
        }
      },
    };
  },
});

File: native/oxlint/rules/three-texture-repeat-requires-wrapping.rs
use oxc_ast::{AstKind, ast::Expression};
use oxc_diagnostics::OxcDiagnostic;
use oxc_semantic::NodeId;
use oxc_span::GetSpan;
use oxc_syntax::operator::AssignmentOperator;

use crate::{
    context::LintContext,
    rule::{Rule, RuleCategory, RuleInfo, RuleMeta},
};

const THREE_TEXTURE_CONSTRUCTOR_NAMES: [&str; 9] = [
    "CanvasTexture",
    "CompressedTexture",
    "Data3DTexture",
    "DataArrayTexture",
    "DataTexture",
    "DepthTexture",
    "FramebufferTexture",
    "Texture",
    "VideoTexture",
];
const THREE_TEXTURE_MODULES: [&str; 3] = ["three", "three-stdlib", "three/"];

#[derive(Debug, Default, Clone)]
pub struct ThreeTextureRepeatRequiresWrapping;

impl RuleMeta for ThreeTextureRepeatRequiresWrapping {
    const NAME: &'static str = "three-texture-repeat-requires-wrapping";
    const PLUGIN: &'static str = "react_doctor_native";
    const CATEGORY: RuleCategory = RuleCategory::Correctness;
    const VERSION: &'static str = "0.1.0";
    const INFO: RuleInfo = RuleInfo {
        short_description: "Require repeat wrapping for repeated Three.js textures.",
    };
}

struct ThreeTextureWrappingFact {
    axis: &'static str,
    node_id: NodeId,
    owner_function_id: Option<NodeId>,
    texture_key: String,
}

impl Rule for ThreeTextureRepeatRequiresWrapping {
    fn run_once<'a>(&self, ctx: &LintContext<'a>) {
        let analysis = build_possible_static_property_write_analysis(ctx);
        let mut wrapping_facts = Vec::new();
        for node in ctx.nodes().iter() {
            match node.kind() {
                AstKind::AssignmentExpression(assignment)
                    if assignment.operator == AssignmentOperator::Assign =>
                {
                    let Some(target) = assignment.left.as_member_expression() else {
                        continue;
                    };
                    let Some(axis) = (match static_member_expression_property_name(target) {
                        Some("wrapS") => Some("x"),
                        Some("wrapT") => Some("y"),
                        _ => None,
                    }) else {
                        continue;
                    };
                    if !three_texture_is_repeating_wrapping(&assignment.right, &analysis, ctx) {
                        continue;
                    }
                    let Some(texture_key) =
                        resolve_expression_key(target.object(), ctx, &mut Vec::new())
                    else {
                        continue;
                    };
                    wrapping_facts.push(ThreeTextureWrappingFact {
                        axis,
                        node_id: node.id(),
                        owner_function_id: local_callback_nearest_function_id(node.id(), ctx),
                        texture_key,
                    });
                }
                AstKind::CallExpression(call_expression) => {
                    let Some(callee) = call_expression
                        .callee
                        .get_inner_expression()
                        .as_member_expression()
                    else {
                        continue;
                    };
                    if static_member_expression_property_name(callee) != Some("set") {
                        continue;
                    }
                    let Some(repeat_member) = callee
                        .object()
                        .get_inner_expression()
                        .as_member_expression()
                    else {
                        continue;
                    };
                    if static_member_expression_property_name(repeat_member) != Some("repeat")
                        || three_constructor_name(
                            repeat_member.object(),
                            &THREE_TEXTURE_CONSTRUCTOR_NAMES,
                            ctx,
                        )
                        .is_none()
                    {
                        continue;
                    }
                    let Some(texture_key) =
                        resolve_expression_key(repeat_member.object(), ctx, &mut Vec::new())
                    else {
                        continue;
                    };
                    let owner_function_id = local_callback_nearest_function_id(node.id(), ctx);
                    for (axis, argument_index, wrapping_name) in
                        [("x", 0, "wrapS"), ("y", 1, "wrapT")]
                    {
                        let Some(expression) = call_expression
                            .arguments
                            .get(argument_index)
                            .and_then(oxc_ast::ast::Argument::as_expression)
                        else {
                            continue;
                        };
                        if resolve_static_number(expression, ctx).is_none_or(|value| value <= 1.0) {
                            continue;
                        }
                        let has_wrapping = wrapping_facts.iter().any(|fact| {
                            fact.axis == axis
                                && fact.texture_key == texture_key
                                && fact.owner_function_id == owner_function_id
                                && ctx.nodes().get_node(fact.node_id).span().start
                                    < node.span().start
                        });
                        if has_wrapping {
                            continue;
                        }
                        ctx.diagnostic(
                            OxcDiagnostic::warn(format!(
                                "texture.repeat.{axis} is greater than one, but the corresponding {wrapping_name} remains ClampToEdgeWrapping so the texture will not tile on that axis"
                            ))
                            .with_label(expression.span()),
                        );
                    }
                }
                _ => {}
            }
        }
    }
}

fn three_texture_is_repeating_wrapping<'a>(
    expression: &Expression<'a>,
    analysis: &PossibleStaticPropertyWriteAnalysis,
    ctx: &LintContext<'a>,
) -> bool {
    ["MirroredRepeatWrapping", "RepeatWrapping"]
        .into_iter()
        .any(|wrapping_name| {
            module_api_reference_matches(
                expression,
                wrapping_name,
                &THREE_TEXTURE_MODULES,
                analysis,
                ctx,
            ) || type_import_module_api_reference_matches(
                expression,
                wrapping_name,
                &THREE_TEXTURE_MODULES,
                analysis,
                ctx,
            )
        })
}
